### ISSUE TITLE
new format

### ISSUE THREAD

**Issue by ziggy2031** [kurupted/google-maps-timeline-viewer#1]
You don't get semantic location history any more. You get timeline.json. I wonder if there is an update coming

**Comment by kurupted** [comment]
Yeah, I'll update as soon as I can, but I'm putting off opting into the new format until I have to.

Once I understand how it works moving forward, and have a sample file to work from, I'll incorporate that.

Can you help with the following?:

1. Since it's on-device, how do you get the file? Some sort of export feature in the app, or is it still via Google Takeout somehow?
2. Have you opened the new and old files in a text editor and confirmed that the format is different?
3. I'd ask for your file but I'm sure you don't want to share it 🙃

**Comment by ziggy2031** [comment]
It's: Settings -> Location -> Location Services -> Timeline -> Export timeline data

**Comment by kurupted** [comment]
Thanks. Seems I don't have that option yet since I'm not opted in.

**Comment by brunoganz18** [comment]
Hi, first of all thank you for this project! I was wondering if you had any updates regarding what @ziggy2031 asked.
Thanks in advance!

**Comment by kurupted** [comment]
I switched to the new system just now, and having an issue finding out how to download the on-device timeline data. When I go to the screen that @ziggy2031 mentioned, I get this:

![Screenshot_20241203-151858](https://github.com/user-attachments/assets/fe397448-d66e-4a84-bfe3-3e415a617f2b)

the export option is on that screen, but there's no way for me to stay on that screen. If I choose my Google account then I no longer see an export option anywhere. 

It may be because I turned on the option to share my location data with Google -- maybe it still gets recorded in the old way basically, and would be exported the old way.

Still looking into it. Interested to hear how it's working for others.

**Comment by kurupted** [comment]
I found that there's this now:

![image](https://github.com/user-attachments/assets/2a1d2073-b9fb-4055-bf56-ed58a22b1e2c)

From here you can turn on syncing from device to cloud, and I think it then gets saved same as before. Will confirm later.

**Comment by brunoganz18** [comment]
@kurupted 
Hi, I have the backup (the one shown in your last attached picture) enabled and when I go into Settings -> Location -> Location Services -> Timeline (I choose my account with the backup enabled) -> I can see the option Export timeline data which generates the .json file.

**Comment by incretoto** [comment]
@kurupted 
I can share my exported timeline data so you can work on it. How do I send it to you?

**Comment by kurupted** [comment]
Thanks, but I was able to export mine and get it working. The format is a bit different so I'm making the code handle both the old and new.

As of right now, when I plot the lines onto the map, it's really rough and missing a lot of granularity. So I need to fix that.

Second also, the new format does not contain place names for your visits, only place IDs. So I'll need to have it query the Maps API for each place, to obtain the name and category

A bit busy atm but will try to get to it soon

**Comment by kurupted** [comment]
I've released the support for exported on-device data files. The walking/driving/etc paths appear to be less detailed, but it may be an issue with the way I'm processing it. I'll look into it further (probably in a week or two).

**Comment by epk** [comment]
Here's my fixes and enhancements to work with the JSON file generated by Google Maps on iOS: 

https://gist.github.com/epk/a70dd9b7a2d5bf8e5d86ebdcaefb6b32

**Comment by kurupted** [comment]
@epk cool. I would imagine the json file format is the same across mobile platforms. Not sure though.

I can't get your project to load. I'm not sure if it's the reason, but the timezone-lookup.min.js url is dead. I can't find a replacement.
The error I'm getting is
Error loading location history: TypeError: data.map is not a function
loadLocationHistory http://localhost/mapv/timeline-viewer.js:324

btw your styles.css file is named syles.css

Also btw, most of my code was Claude slop too ;)

**Comment by kurupted** [comment]
perhaps the reason is that the file formats are indeed different

**Comment by epk** [comment]
@kurupted that error is from the JSON parsing code (and thanks I fixed the CSS typo):
```js
async function loadLocationHistory() {
  try {
    const response = await fetch("location-history.json");
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    const data = await response.json();

    timelineData = data
      .map(item => {
        if (item.visit) {
          const [lat, lng] = item.visit.topCandidate.placeLocation
            .replace("geo:", "")
            .split(",")
            .map(coord => parseFloat(coord));
          return {
            type: "visit",
            name: item.visit.topCandidate.semanticType || "Unknown Location",
            start_time: item.startTime,
            end_time: item.endTime,
            latitude: lat,
            longitude: lng
          };
        } else if (item.activity) {
          const [startLat, startLng] = item.activity.start
            .replace("geo:", "")
            .split(",")
            .map(coord => parseFloat(coord));
          const [endLat, endLng] = item.activity.end
            .replace("geo:", "")
            .split(",")
            .map(coord => parseFloat(coord));
          return {
            type: "activity",
            activity: item.activity.topCandidate.type || "Unknown Activity",
            start_time: item.startTime,
            end_time: item.endTime,
            start_latitude: startLat,
            start_longitude: startLng,
            end_latitude: endLat,
            end_longitude: endLng
          };
        }
        return null;
      })
      .filter(Boolean);

    const selectedDate = moment(document.getElementById("datePicker").value);
    await loadTimelineDataForDate(selectedDate);
  } catch (error) {
    console.error("Error loading location history:", error);
  }
}
```
---
My JSON file from iOS has the following structure:
```
jq -r 'paths(scalars) | map(select(type == "string")) | join(".")' location-history.json | sort -u

activity.distanceMeters
activity.end
activity.probability
activity.start
activity.topCandidate.probability
activity.topCandidate.type
endTime
startTime
timelineMemory.destinations.identifier
timelineMemory.distanceFromOriginKms
timelinePath.durationMinutesOffsetFromStartTime
timelinePath.point
visit.hierarchyLevel
visit.probability
visit.topCandidate.placeID
visit.topCandidate.placeLocation
visit.topCandidate.probability
visit.topCandidate.semanticType
```

**Comment by epk** [comment]
https://github.com/Freika/dawarich/blob/master/app/services/google_maps/phone_takeout_parser.rb has good coverage of all formats

**Comment by cntrlwiz** [comment]
> I've released the support for exported on-device data files. The walking/driving/etc paths appear to be less detailed, but it may be an issue with the way I'm processing it. I'll look into it further (probably in a week or two).

I've fixed the simplifiedRawPath creation.  The path points aren't contained in the semanticSegments activity entries, they are in separate timelinePath entries.

I had to create an array of all of the timelinePath points in the Timeline.json file, then when parsing an activity, I used the start/end timestamps for find and add the relevant points to simplifiedRawPath.

Now the activity paths are properly detailed.

I'm inexperienced with GitHub, as I only code when I need something myself.
I forked your project and did a commit for the changes.
Would you like me to create a pull request from my fork? Or is there some other way I can share the new code?

**Comment by kurupted** [comment]
@epk thanks. My json file from Android looks the same.

For visit, activity, and timelineMemory, I also have
"startTimeTimezoneUtcOffsetMinutes": -300,
"endTimeTimezoneUtcOffsetMinutes": -300,

timelinePath does not have those.

I don't know what the timelineMemory is for. The parser you linked to doesn't deal with them. Probably safe to ignore.

@cntrlwiz that's awesome. I'll take a look shortly. Yeah a pull request would be cool. I also only just code for myself so I'm not too familiar with some of the git features, but I think I just have to approve it.

I also want to fix up the css as per epk's version.

**Comment by kurupted** [comment]
@cntrlwiz I've merged in your code. Thanks!

**Comment by cntrlwiz** [comment]
> @epk thanks. My json file from Android looks the same, so I'm not sure why I can't run your version.
> 
> For visit, activity, and timelineMemory, I also have "startTimeTimezoneUtcOffsetMinutes": -300, "endTimeTimezoneUtcOffsetMinutes": -300,
> 
> timelinePath does not have those.
> 
> I don't know what the timelineMemory is for. The parser you linked to doesn't deal with them. Probably safe to ignore.

The timelineMemory entries appear to be the source for when Timeline would call your travels a "trip" and list the notable places.
I guess the next question is -- do we care to duplicate that function?

**Comment by kurupted** [comment]
ah. I'll add it to the to-do list. 👍
